psbt/psbt.go

```diff
@@ -0,0 +1,957 @@
+// in both the 'Global' and the 'Input' section of a PSBT.
+type Unknown struct {
+	Key   []byte
+	Value []byte
+}
+
+// PInput is a struct encapsulating all the data that can be attached
+// to any specific input of the PSBT.
+type PInput struct {
+	NonWitnessUtxo     *wire.MsgTx
+	WitnessUtxo        *wire.TxOut
+	PartialSigs        []*PartialSig
+	SighashType        txscript.SigHashType
+	RedeemScript       []byte
+	WitnessScript      []byte
+	Bip32Derivation    []*Bip32Derivation
+	FinalScriptSig     []byte
+	FinalScriptWitness []byte
+	Unknowns           []*Unknown
+}
+
+// NewPsbtInput creates an instance of PsbtInput given either a
+// nonWitnessUtxo or a witnessUtxo.
+// NOTE only one of the two arguments should be specified, with the other
+// being `nil`; otherwise the created PsbtInput object will fail IsSane()
+// checks and will not be usable.
+func NewPsbtInput(nonWitnessUtxo *wire.MsgTx,
+	witnessUtxo *wire.TxOut) *PInput {
+	return &PInput{
+		NonWitnessUtxo:     nonWitnessUtxo,
+		WitnessUtxo:        witnessUtxo,
+		PartialSigs:        []*PartialSig{},
+		SighashType:        0,
+		RedeemScript:       nil,
+		WitnessScript:      nil,
+		Bip32Derivation:    []*Bip32Derivation{},
+		FinalScriptSig:     nil,
+		FinalScriptWitness: nil,
+		Unknowns:           nil,
+	}
+}
+
+// IsSane returns true only if there are no conflicting
+// values in the Psbt PInput. It checks that witness and non-witness
+// utxo entries do not both exist, and that witnessScript entries are only
+// added to witness inputs.
+func (pi *PInput) IsSane() bool {
+
+	if pi.NonWitnessUtxo != nil && pi.WitnessUtxo != nil {
+		return false
+	}
+	if pi.WitnessUtxo == nil && pi.WitnessScript != nil {
+		return false
+	}
+	if pi.WitnessUtxo == nil && pi.FinalScriptWitness != nil {
+		return false
+	}
+
+	return true
+}
+
+func (pi *PInput) deserialize(r io.Reader) error {
+	for {
+		keyint, keydata, err := getKey(r)
+		if err != nil {
+			return err
+		}
+		if keyint == -1 {
+			// Reached separator byte
+			break
+		}
+		value, err := wire.ReadVarBytes(r, 0, MaxPsbtValueLength,
+			"PSBT value")
+		if err != nil {
+			return err
+		}
+
+		switch uint8(keyint) {
+
+		case PsbtInNonWitnessUtxo:
+			if pi.NonWitnessUtxo != nil {
+				return ErrDuplicateKey
+			}
+			if keydata != nil {
+				return ErrInvalidKeydata
+			}
+			tx := wire.NewMsgTx(2)
+			if err := tx.Deserialize(bytes.NewReader(value)); err != nil {
+				return err
+			}
+			pi.NonWitnessUtxo = tx
+
+		case PsbtInWitnessUtxo:
+			if pi.WitnessUtxo != nil {
+				return ErrDuplicateKey
+			}
+			if keydata != nil {
+				return ErrInvalidKeydata
+			}
+			txout, err := readTxOut(value)
+			if err != nil {
+				return err
+			}
+			pi.WitnessUtxo = txout
+
+		case PsbtInPartialSig:
+			newPartialSig := PartialSig{PubKey: keydata,
+				Signature: value}
+			if !newPartialSig.checkValid() {
+				return ErrInvalidPsbtFormat
+			}
+			pi.PartialSigs = append(pi.PartialSigs, &newPartialSig)
```

Comment:
https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#handling-duplicated-keys

I think this should check that the public key hasn't already added a signature